dataclients/kubernetes/ingress.go

```diff
@@ -0,0 +1,775 @@
+func (ing *ingress) addSpecRule(ic ingressContext, ru *rule) error {
+	if ru.Http == nil {
+		ic.logger.Warn("invalid ingress item: rule missing http definitions")
+		return nil
+	}
+	// it is a regexp, would be better to have exact host, needs to be added in skipper
+	// this wrapping is temporary and escaping is not the right thing to do
+	// currently handled as mandatory
+	host := []string{"^" + strings.Replace(ru.Host, ".", "[.]", -1) + "$"}
+	// update Traffic field for each backend
+	computeBackendWeights(ic.backendWeights, ru)
+	for _, prule := range ru.Http.Paths {
+		addExtraRoutes(ic, host, ru.Host, prule.Path)
+		if prule.Backend.Traffic > 0 {
+			err := ing.addEndpointsRule(ic, ru.Host, prule)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// converts the default backend if any
+func (ing *ingress) convertDefaultBackend(state *clusterState, i *ingressItem) (*eskip.Route, bool, error) {
+	// the usage of the default backend depends on what we want
+	// we can generate a hostname out of it based on shared rules
+	// and instructions in annotations, if there are no rules defined
+
+	// this is a flaw in the ingress API design, because it is not on the hosts' level, but the spec
+	// tells to match if no rule matches. This means that there is no matching rule on this ingress
+	// and if there are multiple ingress items, then there is a race between them.
+	if i.Spec.DefaultBackend == nil {
+		return nil, false, nil
+	}
+
+	var (
+		eps     []string
+		err     error
+		ns      = i.Metadata.Namespace
+		name    = i.Metadata.Name
+		svcName = i.Spec.DefaultBackend.ServiceName
+		svcPort = i.Spec.DefaultBackend.ServicePort
+	)
+
+	svc, err := state.getService(ns, svcName)
+	if err != nil {
+		log.Errorf("convertDefaultBackend: Failed to get service %s, %s, %s", ns, svcName, svcPort)
+		return nil, false, err
+	}
+
+	targetPort, err := svc.getTargetPort(svcPort)
+	if err != nil {
+		err = nil
+		log.Errorf("Failed to find target port %v, %s, fallback to service", svc.Spec.Ports, svcPort)
+	} else if svc.Spec.Type == "ExternalName" {
+		scheme := "https"
+		if targetPort != "443" {
+			scheme = "http"
+		}
+		u := fmt.Sprintf("%s://%s:%s", scheme, svc.Spec.ExternalName, targetPort)
+		f, err := eskip.ParseFilters(fmt.Sprintf(`setRequestHeader("Host", "%s")`, svc.Spec.ExternalName))
+		if err != nil {
+			return nil, false, err
+		}
+		return &eskip.Route{
+			Id:          routeID(ns, name, "default", "", svc.Spec.ExternalName),
+			BackendType: eskip.NetworkBackend,
+			Backend:     u,
+			Filters:     f,
+		}, true, nil
+	} else {
+		log.Debugf("Found target port %v, for service %s", targetPort, svcName)
+		eps, err = state.getEndpoints(
+			ns,
+			svcName,
+			svcPort.String(),
+			targetPort,
+		)
+		log.Debugf("convertDefaultBackend: Found %d endpoints for %s: %v", len(eps), svcName, err)
+	}
+
+	if len(eps) == 0 || err == errEndpointNotFound {
```

Comment:
Based on our discussion: do we want to drop this service handling if we have a service type ClusterIP?

`svc.Spec.Type == "ClusterIP"`

see also above "ExternalName" https://github.com/zalando/skipper/pull/1180/files#diff-cc26174bd99fb74da18dffde42bd2efcR502

Reply:
yes, on it

Reply:
i mean do we want to change this behavior in this PR?

Reply:
yes

Reply:
we have to postpone this, the changes are stored in this branch:

https://github.com/zalando/skipper/tree/improvement/service-port-fallback-handling

the majority of the kubernetes tests are not written with endpoints and target ports in mind. This means that a huge chunk of the tests would need to be changed, which makes the whole PR quite unsafe